resources: Add source code for riscv-tests

kunpai opened the pull request:
- This code is compatible with m5ops, i.e., ROIs are set using m5_work_begin and m5_work_end.
- Added README.md and CHANGELOG.md for every benchmark to track issues
- Updated base README.md of asmtest directory to show what code has been changed
- riscv-tests benchmarks' source code is now housed inside src/asmtest/benchmarks. Earlier, it used to be in src/riscv-tests

Change-Id: I17572d00b50617bc52e8453ee873b0c2f614f461

kunpai committed:
resources: Move riscv-tests/benchmarks into asmtest

Since riscv-tests is outdated, and its benchmarks
are to be housed in asmtest, this patch moves
benchmarks into asmtest and deletes riscv-tests/.

kunpai committed:
resources: Update asmtest with current riscv-tests

This commit updates asmtest and its files with the current state
of riscv-tests as of 3 Aug 2023.

kunpai committed:
resources: Edit source code of riscv-tests

This patch edits source code of riscv-tests
for gem5 compatibility.

These changes are:

- Add m5_work_begin and m5_work_end instead
of spike setStats.

- Comment out spike magic instructions where
it is unnecessary to core functionality of the
benchmark (for example, Start_Timer(); in
dhrystone).

- Add README.md and CHANGELOG.md for every
benchmark.

- Edit Makefile to not use riscv-tests linkers
and instead just use the common directory and
gem5.

- Move encoding.h into the common directory for
the above reason.

BobbyRBruce commented:
@kunpai: Are all these benchmarks now built and in the database?

kunpai commented:
> @kunpai: Are all these benchmarks now built and in the database?

Hey @BobbyRBruce, I am unsure of the procedure to merge these, so I opened the PR and thought to submit the JSON objects when this is merged. Should I do the opposite? Or both in parallel?

BobbyRBruce commented:
> 



> > @kunpai: Are all these benchmarks now built and in the database?
> 
> Hey @BobbyRBruce, I am unsure of the procedure to merge these, so I opened the PR and thought to submit the JSON objects when this is merged. Should I do the opposite? Or both in parallel?

Didn't we already put the JSON objects in the database though? I put some risc-v benchmarks there.

kunpai commented:
Those were the RISCV microbenchmarks, these are the RISCV-tests binaries

On Wed, Aug 16, 2023 at 10:34 AM Bobby R. Bruce ***@***.***>
wrote:

> @kunpai <https://github.com/kunpai>: Are all these benchmarks now built
> and in the database?
>
> Hey @BobbyRBruce <https://github.com/BobbyRBruce>, I am unsure of the
> procedure to merge these, so I opened the PR and thought to submit the JSON
> objects when this is merged. Should I do the opposite? Or both in parallel?
>
> Didn't we already put the JSON objects in the database though? I put some
> risc-v benchmarks there.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/gem5/gem5-resources/pull/4#issuecomment-1681018134>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APAPZ6DBC42C6VGOT4RG4G3XVUACZANCNFSM6AAAAAA3ENKKGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
-- 
Regards,

Kunal S. Pai

BobbyRBruce commented:
> Those were the RISCV microbenchmarks, these are the RISCV-tests binaries
> On Wed, Aug 16, 2023 at 10:34 AM Bobby R. Bruce ***@***.***> wrote: @kunpai <https://github.com/kunpai>: Are all these benchmarks now built and in the database? Hey @BobbyRBruce <https://github.com/BobbyRBruce>, I am unsure of the procedure to merge these, so I opened the PR and thought to submit the JSON objects when this is merged. Should I do the opposite? Or both in parallel? Didn't we already put the JSON objects in the database though? I put some risc-v benchmarks there. — Reply to this email directly, view it on GitHub <[#4 (comment)](https://github.com/gem5/gem5-resources/pull/4#issuecomment-1681018134)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/APAPZ6DBC42C6VGOT4RG4G3XVUACZANCNFSM6AAAAAA3ENKKGQ> . You are receiving this because you were mentioned.Message ID: ***@***.***>
> -- Regards, Kunal S. Pai

1. So, did we add sources for these tests? If not, shouldn't we?
2. Ignoring 1., can you send me the resources and JSON to add to gem5-resources? I think we should do these in parallel.

rogerchang23424 commented:
Something I would like to mention in the PR

1. I have do some update in asmtest resources before, so the version in develop branch is newer than stable version. The asmtest in develop branch is [394937eceb62bc4ba8ad5a4e916f0508d9803229](https://github.com/riscv-software-src/riscv-tests/tree/394937eceb62bc4ba8ad5a4e916f0508d9803229)
2. The riscv-tests folder is older than asmtest in develop branch, which is based on [e65ecdf941a5484af27f9be223fb655ebcb0398b](https://github.com/riscv-software-src/riscv-tests/tree/e65ecdf941a5484af27f9be223fb655ebcb0398b). It should be updated before merging to the asmtest.

Feel free to ping me if you have any concern, thanks.

kunpai commented:
> Something I would like to mention in the PR
> 
> 1. I have do some update in asmtest resources before, so the version in develop branch is newer than stable version. The asmtest in develop branch is [394937eceb62bc4ba8ad5a4e916f0508d9803229](https://github.com/riscv-software-src/riscv-tests/tree/394937eceb62bc4ba8ad5a4e916f0508d9803229)
> 2. The riscv-tests folder is older than asmtest in develop branch, which is based on [e65ecdf941a5484af27f9be223fb655ebcb0398b](https://github.com/riscv-software-src/riscv-tests/tree/e65ecdf941a5484af27f9be223fb655ebcb0398b). It should be updated before merging to the asmtest.
> 
> Feel free to ping me if you have any concern, thanks.

You are right, the riscv-tests folder is older than asmtest. Since the riscv-tests benchmarks are part of asmtest, we decided to move the source code into asmtest. So, we are deleting the outdated files (but keeping the files unique to gem5-resources), pulling the last version into asmtest, and then making changes like adding m5ops, to improve compatibility with gem5.

kunpai commented:
> > Those were the RISCV microbenchmarks, these are the RISCV-tests binaries
> > On Wed, Aug 16, 2023 at 10:34 AM Bobby R. Bruce _**@**_._**> wrote: @kunpai https://github.com/kunpai: Are all these benchmarks now built and in the database? Hey @BobbyRBruce https://github.com/BobbyRBruce, I am unsure of the procedure to merge these, so I opened the PR and thought to submit the JSON objects when this is merged. Should I do the opposite? Or both in parallel? Didn't we already put the JSON objects in the database though? I put some risc-v benchmarks there. — Reply to this email directly, view it on GitHub <[#4 (comment)](https://github.com/gem5/gem5-resources/pull/4#issuecomment-1681018134)>, or unsubscribe https://github.com/notifications/unsubscribe-auth/APAPZ6DBC42C6VGOT4RG4G3XVUACZANCNFSM6AAAAAA3ENKKGQ . You are receiving this because you were mentioned.Message ID: _**@**_.**_>
> > -- Regards, Kunal S. Pai
> 
> 1. So, did we add sources for these tests? If not, shouldn't we?
> 2. Ignoring 1., can you send me the resources and JSON to add to gem5-resources? I think we should do these in parallel.

1. I think we should, since the source code for gem5 is different from the riscv-tests source code. I added m5ops and removed some magic instructions and print statements that were causing problems during execution.
2. Sure, I will work on the JSON objects for each benchmark.

rogerchang23424 commented on src/asmtest/env/p/riscv_test.h:
```diff
@@ -188,23 +198,13 @@ handle_exception:                                                       \
         /* some unhandlable exception occurred */                       \
   1:    ori TESTNUM, TESTNUM, 1337;                                     \
   write_tohost:                                                         \
-        mv a1, TESTNUM;                                                 \
-        /* 1 = success; otherwise (a0 >> 1) is the failed test */       \
-        srl a1, a1, 1;                                                  \
-        la a0, tohost_data;                                             \
-        sb a1, 0(a0);                                                   \
-        /* write exit code (binary) to file */                          \
-        li a1, 1;                                                       \
-        li a2, 0;                                                       \
-        la a3, tohost_file;                                             \
-        .long 0x9E00007B;                                               \
-        /* shutdown gem5 */                                             \
-        li a0, 0;                                                       \
-        .long 0x4200007B;                                               \
```
The section cloud have different implementation due to each simulator has their own HTIF. The section is m5ops to write test results to host and exit simulation in RISC-V bare metal(FS) mode. Any reason to change back?

rogerchang23424 commented on src/asmtest/env/v/vm.c:
```diff
@@ -17,47 +17,47 @@
-static const char *stdoutFile = "stdout";
-static const char *exitcodeFile = "exitcode";
-
 static uint64_t lfsr63(uint64_t x)
 {
   uint64_t bit = (x ^ (x >> 1)) & 1;
   return (x >> 1) | (bit << 62);
 }
 
-static void cputchar(const char *file, int x)
+static void cputchar(int x)
 {
-  volatile char c = x;
-  register size_t a0 asm("a0") = (uintptr_t)(&c);
-  register size_t a1 asm("a1") = 1;
-  register size_t a2 asm("a2") = 0;
-  register size_t a3 asm("a3") = (uintptr_t)file;
-  asm volatile (".long 0x9E00007B" : : "r"(a0), "r"(a1), "r"(a2), "r"(a3));
+  do_tohost(0x0101000000000000 | (unsigned char)x);
 }
 
 static void cputstring(const char* s)
 {
   while (*s)
-    cputchar(stdoutFile, *s++);
+    cputchar(*s++);
 }
 
 static void terminate(int code)
 {
-  cputchar(exitcodeFile, code >> 1);
-  register size_t a0 asm("a0") = 0;
-  asm volatile (".long 0x4200007B" : : "r"(a0));
+  do_tohost(code);
   while (1);
 }
 
 void wtf()
 {
-  terminate(255 << 1);
```
The section is the m5ops in v mode. Should it need to change back?

rogerchang23424 commented on src/asmtest/Makefile:
```diff
@@ -33,13 +34,15 @@ include $(src_dir)/rv32uamt/Makefrag
 #--------------------------------------------------------------------
 
-RISCV_PREFIX ?= riscv$(XLEN)-unknown-elf-
+RISCV_PREFIX ?= riscv$(XLEN)-linux-gnu-
```
Any reason to change to `linux-gnu`?

rogerchang23424 reviewed (commented):
> > Something I would like to mention in the PR
> > 
> > 1. I have do some update in asmtest resources before, so the version in develop branch is newer than stable version. The asmtest in develop branch is [394937eceb62bc4ba8ad5a4e916f0508d9803229](https://github.com/riscv-software-src/riscv-tests/tree/394937eceb62bc4ba8ad5a4e916f0508d9803229)
> > 2. The riscv-tests folder is older than asmtest in develop branch, which is based on [e65ecdf941a5484af27f9be223fb655ebcb0398b](https://github.com/riscv-software-src/riscv-tests/tree/e65ecdf941a5484af27f9be223fb655ebcb0398b). It should be updated before merging to the asmtest.
> > 
> > Feel free to ping me if you have any concern, thanks.
> 
> You are right, the riscv-tests folder is older than asmtest. Since the riscv-tests benchmarks are part of asmtest, we decided to move the source code into asmtest. So, we are deleting the outdated files (but keeping the files unique to gem5-resources), pulling the last version into asmtest, and then making changes like adding m5ops, to improve compatibility with gem5.

Look good. I have compared the PR with riscv-tests, I find some files are still using the old implementation, any reason for it?

Thanks

kunpai commented:
> > > Something I would like to mention in the PR
> > > 
> > > 1. I have do some update in asmtest resources before, so the version in develop branch is newer than stable version. The asmtest in develop branch is [394937eceb62bc4ba8ad5a4e916f0508d9803229](https://github.com/riscv-software-src/riscv-tests/tree/394937eceb62bc4ba8ad5a4e916f0508d9803229)
> > > 2. The riscv-tests folder is older than asmtest in develop branch, which is based on [e65ecdf941a5484af27f9be223fb655ebcb0398b](https://github.com/riscv-software-src/riscv-tests/tree/e65ecdf941a5484af27f9be223fb655ebcb0398b). It should be updated before merging to the asmtest.
> > > 
> > > Feel free to ping me if you have any concern, thanks.
> > 
> > 
> > You are right, the riscv-tests folder is older than asmtest. Since the riscv-tests benchmarks are part of asmtest, we decided to move the source code into asmtest. So, we are deleting the outdated files (but keeping the files unique to gem5-resources), pulling the last version into asmtest, and then making changes like adding m5ops, to improve compatibility with gem5.
> 
> Look good. I have compared the PR with riscv-tests, I find some files are still using the old implementation, any reason for it?
> 
> Thanks

I saw your comments, and it seems like these files were missed out. Thanks for letting me know, I will be sure to push a patch to it from its version in `develop`

rogerchang23424 commented on src/asmtest/benchmarks/Makefile:
```diff
@@ -23,21 +23,27 @@ bmarks = \
-	mt-vvadd \
-	mt-matmul \
 	pmp \
+	# vec-memcpy \
+	# vec-daxpy \
+	# vec-sgemm \
```
(nit) It might work after gem5 support vector instructions

kunpai replied:
Yeah, when it does, I think we should push a patch for this directory

rogerchang23424 commented on src/asmtest/benchmarks/Makefile:
```diff
@@ -23,21 +23,27 @@ bmarks = \
 #--------------------------------------------------------------------
 
-RISCV_PREFIX ?= riscv$(XLEN)-unknown-elf-
+RISCV_PREFIX ?= riscv$(XLEN)-linux-gnu-
```
Just curious why to change to `linux-gnu`?

rogerchang23424 reviewed (commented):
This PR is LGTM. Cloud you move the PR target to `develop` branch? It will be easier to compare the `riscv-tests` version.

rogerchang23424 reviewed (changes requested):
Sorry, I clicked too fast.

The PR merge conflict should be solved

kunpai commented:
Hi Roger, I have fixed the merge conflict on my end by combining both the READMEs present. I believe only someone with write access to the repository can resolve the conflict. I might be mistaken, though.

Tagging @BobbyRBruce for further direction on this PR

rogerchang23424 commented:
Hi @kunpai,

You should rebase the PR commits on the top of develop branch and resolve the merge conflict by youself, thanks.

kunpai commented:
> Hi @kunpai,
> 
> You should rebase the PR commits on the top of develop branch and resolve the merge conflict by youself, thanks.

Done. Thanks for the direction, that helped!

rogerchang23424 reviewed (approved):
LGTM
